pennylane/labs/trotter/realspace/matrix.py

```diff
@@ -0,0 +1,177 @@
+    matrix[rows, cols] = vals
+
+    return 1 / np.sqrt(2) * matrix
+
+
+def _momentum_operator(
+    gridpoints: int, sparse: bool = False, basis: str = "realspace"
+) -> Union[np.ndarray, sp.sparse.csr_array]:
+    """Returns a matrix representation of the momentum operator"""
+
+    if basis == "realspace":
+        matrix = _realspace_momentum(gridpoints)
+        return sp.sparse.csr_array(matrix) if sparse else matrix
+
+    if basis == "harmonic":
+        matrix = _harmonic_momentum(gridpoints)
+        return sp.sparse.csr_array(matrix) if sparse else matrix
+
+    raise ValueError(f'"{basis}" is not a valid basis')
+
+
+def _realspace_momentum(gridpoints: int) -> np.ndarray:
+    """Returns a matrix representation of the momenumt operator in the realspace basis"""
+    values = np.arange(gridpoints)
+    values[gridpoints // 2 :] -= gridpoints
+    values = values * (np.sqrt(2 * np.pi / gridpoints))
+    dft = sp.linalg.dft(gridpoints, scale="sqrtn")
+    matrix = dft @ np.diag(values) @ dft.conj().T
+
+    return matrix
+
+
+def _harmonic_momentum(gridpoints: int) -> np.ndarray:
+    """Returns a matrix representation of the momenumt operator in the harmonic basis"""
+    rows = np.array(list(range(1, gridpoints)) + list(range(0, gridpoints - 1)))
+    cols = np.array(list(range(0, gridpoints - 1)) + list(range(1, gridpoints)))
+    vals = np.array(
+        [np.sqrt(i) for i in range(1, gridpoints)] + [-np.sqrt(i) for i in range(1, gridpoints)]
+    )
+
+    matrix = np.zeros(shape=(gridpoints, gridpoints))
+    matrix[rows, cols] = vals
+
+    return (1j / np.sqrt(2)) * matrix
+
+
+def _string_to_matrix(
+    op: str, gridpoints: int, sparse: bool = False, basis: str = "realspace"
+) -> Union[np.ndarray, sp.sparse.csr_array]:
+    """Transforms a string of P's and Q's into a matrix representing the product of position and momenutm operators"""
+
+    matrix = _identity(gridpoints, sparse=sparse)
+    p = _momentum_operator(gridpoints, basis=basis, sparse=sparse)
+    q = _position_operator(gridpoints, basis=basis, sparse=sparse)
+
+    for char in op:
+        if char == "P":
+            matrix = matrix @ p
+            continue
+
+        if char == "Q":
+            matrix = matrix @ q
+            continue
+
+        raise ValueError(f"Operator terms must only contain P and Q. Got {char}.")
+
+    return matrix
+
+
+def _tensor_with_identity(
+    modes: int,
+    gridpoints: int,
+    index: Tuple[int],
+    ops: Tuple[Union[np.ndarray, sp.sparse.csr_array]],
+    sparse: bool = False,
+) -> Union[np.ndarray, sp.sparse.csr_array]:
+    """Tensor the input matrices with the identity"""
+    lookup = {}
+
+    for mode in range(modes):
+        lookup[mode] = _identity(gridpoints, sparse=sparse)
+        for count, i in enumerate(index):
+            if i == mode:
+                lookup[mode] = lookup[mode] @ ops[count]
+
+    matrix = lookup[0]
+    for mode in range(1, modes):
+        matrix = _kron(matrix, lookup[mode])
```

Comment:
```suggestion
    matrix = reduce(lambda x, y: _kron(x, y), lookup[1:], lookup[0])
```
this might be more efficient